### DISCUSSION TITLE
Is issue for ToggleButtonAssist.SwitchTrackOnBackground of ToggleButton?

### DISCUSSION THREAD

**Discussion by guytechNet** [MaterialDesignInXAML/MaterialDesignInXamlToolkit#3338] [General]
Hi All,

What's color? I think: It is not green or red color inside **ToggleButtonAssist.SwitchTrackOnBackground of ToggleButton**. 

![Screenshot 2023-10-13 091324](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/assets/118602474/97243836-0d94-45ad-b418-d02616dca25c)

![Screenshot 2023-10-13 091003](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/assets/118602474/ae572e1a-9a21-4847-ad34-e1b8c9b08198)

**Please, Do you review for me?**

**Comment by nicolaihenriksen** [comment]
@guytechNet If you inspect it with Snoop, you will see that those colors are in fact set as the `Fill` on the "track" of the toggle. However, the colors are then dimmed down by the style hardcoding `Opacity=0.26` (see [here](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/blob/29bcb91d6c68203d9515bd659c4d6f16ff35b4a5/MaterialDesignThemes.Wpf/Themes/MaterialDesignTheme.ToggleButton.xaml#L441))

Whether that is desirable/intentional I cannot really comment on. Perhaps @Keboo could have a look? Internally the template sets the `ToggleButtonAssist.SwitchTrachOnBackground` (and the "off" variant), so determining if the value is the default or set from outside may be difficult. Perhaps to allow better control, the `Opacity` for the on/off states could be exposed as an attached property?


![image](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/assets/19572699/e57a7aee-f49e-4256-8abf-5e1a27c26c93)

![image](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/assets/19572699/8bbb8028-fd10-4147-b5c2-367809c9550b)

![image](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/assets/19572699/9c122835-3219-4661-96e0-6a9a002640af)

**Reply by guytechNet** [reply]
Thanks your answer! @Keboo please take care it for me! Can you fix it?

**Reply by Keboo** [reply]
Yea the [hard coded Opacity](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/blob/master/MaterialDesignThemes.Wpf/Themes/MaterialDesignTheme.ToggleButton.xaml#L441) is what is doing it. I suspect what we want to do instead is have the default brushes have an alpha channel defined, and if someone explicitly sets a color via the attached properties just use it directly. Then we can remove the hard coded Opacity levels.

**Reply by nicolaihenriksen** [reply]
@Keboo As a general comment I actually prefer the use of alpha-channels on brushes over using opacity on elements. Simply because I have too often seen the `IsEnabled=False` being used as a trigger to set the opacity on elements. The unfortunate thing about that is that `IsEnabled` is an inheritable property, and thus if you have the same trigger approach to setting an opacity for nested elements, you can end up with "double" (or multiple) opacity unless you are very strict about where you apply the opacities. This can end up rendering some elements nearly invisible if the nesting is deep enough.

However, if we do go down the alpha channel path, won't that require us to generate multiple brushes for the various desired opacities? Today the background brush of the track is set like this:
```xaml
<Setter Property="wpf:ToggleButtonAssist.SwitchTrackOnBackground" Value="{DynamicResource MaterialDesign.Brush.Primary.Light}" />
```

And I guess you don't want `MaterialDesign.Brush.Primary.Light` to have an alpha channel matching the opacity?! I would assume you would need something like:

`MaterialDesign.Brush.Primary.Light.DimmedBackground`
`MaterialDesign.Brush.Primary.Light.DisabledState`
`MaterialDesign.Brush.Primary.Light.SomeOtherState`

And this is where I could foresee it run amok because how do we define these states, and can we just randomly add new ones? Also, are these states needed for ALL brushes? Anyway, I am not against your proposal, I just think we need to think it through.

**Reply by Keboo** [reply]
Completely agree on the alpha channels and brushes. However, I _think_ I would actually take the brush approach further. Rather than just using `Primary.Light` I think it would be nice to have control specific brushes. So, in this case brushes that are specific to the switch control.  With the new 5.0 theme stuff, it is not that difficult to gen new these brushes making this a much more viable option.

The other thing to consider is if we actually want the alpha channel, or if we would prefer to instead generate the equivalent color. Meaning assume a given background color and generate the appropriate RGB to get the same effective look for the brush. We already have code to do this color flattening and I have used it on other brushes. Because no matter how do you it, with transparency there is always the potential for an odd look if it gets placed on a weird background color. Not sure that is really the case here, but an option worth considering.

**Reply by guytechNet** [reply]
Hi @Keboo, 
Now, I want to use red and green for ToggleButtonAssist.SwitchTrachOnBackground. Can I do it? How do I do?
Do you upgrade toolkit?

**Reply by nicolaihenriksen** [reply]
@guytechNet For now, until something is added to the library, you have to create a local copy of the `Style` and apply that style on the element in question. I think that is the easiest way for you to get rid of the hardcoded opacity with the current implementation.

**Reply by guytechNet** [reply]
Hi @nicolaihenriksen, 
I think that is mistake for them (as style opacity on the element in my question)
Please, fix it and upgrade toolkit. Thanks!

**Reply by guytechNet** [reply]
Hi @nicolaihenriksen,
Can you help me create a local copy of the Style and apply that style on the element in my question?

**Reply by nicolaihenriksen** [reply]
@guytechNet Sure I can help out.

One option is to simply copy all of the contents of [this style](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/blob/73c546f4078ac0f54947e1976d4e92b74ce5aff0/MaterialDesignThemes.Wpf/Themes/MaterialDesignTheme.ToggleButton.xaml#L348-L531) into your own style like this:
```xaml
<Style x:Key="MyCustomMaterialDesignSwitchToggleButton" TargetType="{x:Type ToggleButton}">
  ... paste the contents of the style linked above in here and modify the Opacity on the element named 'Track' to "1"...
</Style>
```
and then simply apply your style to your toggle button:
```xaml
<ToggleButton Style="{StaticResource MyCustomMaterialDesignSwitchToggleButton}">
 ...
</ToggleButton>
```
This will work, but it also means you opt-out of using the actual style that ships with MDIX.

Another approach could be to attempt to leverage the style that ships with MDIX (and any updates it may receive), but add custom functionality on top; in particularly functionality which removes the opacity set on the track. The approach below uses an attached property which, when set, hooks up to the `Loaded` event of the `ToggleButton` and set the opacity on the child element named "Track". This works, but it also makes an assumption about the internal naming of children within the style; so not a perfect solution. In reality, should probably also be added as an attached behavior rather than an attached property. Anyway, it gets the job done.

You can pick which one you like better.

You apply the attached property in this manner:
```xaml
<ToggleButton
    Style="{StaticResource MaterialDesignSwitchToggleButton}"
    local:ToggleButtonAssistExt.RemoveTrackOpacity="True">
  <materialDesign:ToggleButtonAssist.SwitchTrackOnBackground>
    <SolidColorBrush Color="Green" />
  </materialDesign:ToggleButtonAssist.SwitchTrackOnBackground>

  <materialDesign:ToggleButtonAssist.SwitchTrackOffBackground>
    <SolidColorBrush Color="Red" />
  </materialDesign:ToggleButtonAssist.SwitchTrackOffBackground>
</ToggleButton>
```

The result will look like this:
![ToggleButtonTrackColor](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/assets/19572699/0b134b21-286b-41ca-9d76-9ab44b737610)

The `ToggleButtonAssistExt.cs` looks like this. Note that it does not explicitly handle the case where the attached property is first set and then unset at a later point. I don't think this is a common use case, but worth mentioning that it could be improved upon if needed.
```csharp
public static class ToggleButtonAssistExt
{
    public static readonly DependencyProperty RemoveTrackOpacityProperty = DependencyProperty.RegisterAttached(
        "RemoveTrackOpacity", typeof(bool), typeof(ToggleButtonAssistExt), new PropertyMetadata(default(bool), RemoveTrackOpacityChanged));

    private static void RemoveTrackOpacityChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
    {
        if (d is ToggleButton toggleButton)
        {
            if (Equals(true, e.NewValue))
            {
                SetupRemoveTrackOpacity();
            }
        }

        void SetupRemoveTrackOpacity()
        {
            toggleButton.Loaded -= ToggleButtonOnLoaded;    // Avoid multiple event subscriptions
            toggleButton.Loaded += ToggleButtonOnLoaded;
        }

        void ToggleButtonOnLoaded(object sender, RoutedEventArgs e)
        {
            if (toggleButton.FindChild<Rectangle>("Track") is { } track)
            {
                track.Opacity = GetRemoveTrackOpacity(toggleButton) ? 1 : 0.26;
            }
        }
   }

    public static void SetRemoveTrackOpacity(ToggleButton element, bool value)
        => element.SetValue(RemoveTrackOpacityProperty, value);

    public static bool GetRemoveTrackOpacity(ToggleButton element)
        => (bool) element.GetValue(RemoveTrackOpacityProperty);


    /// <summary>
    /// NOTE: Direct copy paste from TreeHelper.cs of the MDIX library
    /// 
    /// Finds a Child of a given item in the visual tree. 
    /// </summary>
    /// <param name="parent">A direct parent of the queried item.</param>
    /// <typeparam name="T">The type of the queried item.</typeparam>
    /// <param name="childName">x:Name or Name of child. </param>
    /// <returns>The first parent item that matches the submitted type parameter. 
    /// If not matching item can be found, 
    /// a null parent is being returned.</returns>
    public static T? FindChild<T>(this DependencyObject parent, string childName)
        where T : DependencyObject
    {
        // Confirm parent and childName are valid. 
        if (parent == null) return null;

        T? foundChild = null;

        int childrenCount = VisualTreeHelper.GetChildrenCount(parent);
        for (int i = 0; i < childrenCount; i++)
        {
            var child = VisualTreeHelper.GetChild(parent, i);
            // If the child is not of the request child type child
            if (child is not T)
            {
                // recursively drill down the tree
                foundChild = FindChild<T>(child, childName);

                // If the child is found, break so we do not overwrite the found child. 
                if (foundChild != null) break;
            }
            else if (!string.IsNullOrEmpty(childName))
            {
                // If the child's name is set for search
                if (child is FrameworkElement frameworkElement && frameworkElement.Name == childName)
                {
                    // if the child's name is of the request name
                    foundChild = (T)child;
                    break;
                }
            }
            else
            {
                // child element found.
                foundChild = (T)child;
                break;
            }
        }
        return foundChild;
    }
}
```

**Reply by guytechNet** [reply]
Thanks @nicolaihenriksen very much.
I have chosen the ToggleButtonAssistExt.cs class for my issue, and it worked successfully

**Comment by guytechNet** [comment]
Hi @Keboo, @nicolaihenriksen 
I want to use **red background**  and **green background** for **ToggleButtonAssist.SwitchTrachOnBackground**. 
How do I do? Do you upgrade for them? Thanks!!!
